client/rpc/src/state/mod.rs

```diff
@@ -102,8 +102,19 @@ pub trait StateBackend<Block: BlockT, Client>: Send + Sync + 'static
 		block: Option<Block::Hash>,
 		key: StorageKey,
 	) -> FutureResult<Option<u64>> {
-		Box::new(self.storage(block, key)
-			.map(|x| x.map(|x| x.0.len() as u64)))
+		let value_len = self.storage(block, key.clone()).map(|x| x.map(|x| x.0.len() as u64));
+		let pairs_len = self.storage_pairs(block, key.clone())
```

Comment:
Shouldn't that query be done conditionally only if `value_len` results into `None`?

Reply:
yeah that makes more sense and I initially started with that, but failed to get it to work. Ideally I want: only if the first one is `Ok(None)`, then try the second one. 

But I am not sure if this is even possible? I am but new to futures.

Reply:
Depends on the futures version we have here, but it seems it's 0.1 and I think it should be something like this:
```rust
value_len.and_then(|result| match result {
  None => Either::A(self.storage_pairs(...)...),
  Some(v) => Either:B(futures::done(Ok(Some(v))),
})
```

Reply:
I'm pretty sure I tried something like that and failed once, but let me try again and get back.

Reply:
Okay so this was the first time I was working with futures and the main point of confusion was the versioning going on (which makes me appreciate more and more the standardised async/await). I wasn't aware that the `rpc` us using and re-exporting `0.1` and the current crate is using `0.3`, and I was getting strange errors all over the place.

Reply:
The `storage_pairs` function will return the results directly. You need to do this differently here.

Reply:
can you explain more? I don't see the difference in what you suggested.

Reply:
When you call a function that returns a future, you can also directly return the result. This means the first time the future is polled the result is ready. However this also means that the operation is always executed, in this case the scanning of the storage. I propose that you put the scanning call into a closure and only call this closure if the first storage call returned a none. This means you will only scan the storage when the value does not exists vs query it always.

Reply:
I exactly wanted to do as you explain: only query the second if the first one is `None`. This is how I thought: calling `self.storage_pairs` and storing it in `pairs_len` executes nothing, because it is not yet polled. I thought that also afterwards when I pass it to `and_then`, it is would be only polled in the `None` arm, but apparently not.